src/profile-logic/zip-files.js

```diff
@@ -0,0 +1,269 @@
+
+    let path = '';
+    let prefixIndex = null;
+    for (let j = 0; j < pathParts.length; j++) {
+      // Go through each path part to assemble the table
+      const pathPart = pathParts[j];
+
+      // Add the path part to the path.
+      if (path) {
+        path += '/' + pathPart;
+      } else {
+        path = pathPart;
+      }
+
+      // This part of the path may already exist.
+      const existingIndex = pathToFilesTableIndex.get(path);
+      if (existingIndex !== undefined) {
+        // This folder was already added, so skip it, but remember the prefix.
+        prefixIndex = existingIndex;
+        continue;
+      }
+
+      const index = filesTable.length++;
+      filesTable.prefix[index] = prefixIndex;
+      filesTable.path[index] = path;
+      filesTable.partName[index] = pathPart;
+      filesTable.depth[index] = j;
+      filesTable.file[index] =
+        j + 1 === pathParts.length ? zipEntries.files[fullPath] : null;
+      pathToFilesTableIndex.set(path, index);
+      // Remember this index as the prefix.
+      prefixIndex = index;
+    }
+  }
+  return filesTable;
+}
+
+export function getZipFileMaxDepth(zipFileTable: ZipFileTable | null): number {
+  if (!zipFileTable) {
+    return 0;
+  }
+  let maxDepth = 0;
+  for (let i = 0; i < zipFileTable.length; i++) {
+    maxDepth = Math.max(maxDepth, zipFileTable.depth[i]);
+  }
+  return maxDepth;
+}
+
+export class ZipFileTree {
+  _zipFileTable: ZipFileTable;
+  _parentToChildren: null | Map<
+    IndexIntoZipFileTable | null,
+    IndexIntoZipFileTable[]
+  >;
+  _displayDataByIndex: Map<IndexIntoZipFileTable, ZipDisplayData>;
+  _zipFileUrl: string;
+
+  constructor(zipFileTable: ZipFileTable, zipFileUrl: string) {
+    this._zipFileTable = zipFileTable;
+    this._zipFileUrl = zipFileUrl;
+    this._displayDataByIndex = new Map();
+  }
+
+  getRoots(): IndexIntoZipFileTable[] {
+    return this.getChildren(null);
+  }
+
+  getChildren(
+    zipTableIndex: IndexIntoZipFileTable | null
+  ): IndexIntoZipFileTable[] {
+    const parentToChildMap = this._getParentToChildMap();
+    const children = parentToChildMap.get(zipTableIndex);
+    if (!children) {
+      throw new Error(
+        'Attempted to fetch the children from an unknown IndexIntoZipFileTable.'
+      );
+    }
+    return children;
+  }
+
+  _computeChildrenArray(
+    parentIndex: IndexIntoZipFileTable | null
+  ): IndexIntoZipFileTable[] {
+    const children = [];
+    for (
+      let childIndex = 0;
+      childIndex < this._zipFileTable.length;
+      childIndex++
+    ) {
+      if (this._zipFileTable.prefix[childIndex] === parentIndex) {
+        children.push(childIndex);
+      }
+    }
+    return children;
+  }
+
+  /**
+   * Create a Map of the parents to the children to make it O(1) to dynamically compute
+   * any property about the tree.
+   */
+  _getParentToChildMap(): Map<
+    IndexIntoZipFileTable | null,
+    IndexIntoZipFileTable[]
+  > {
```

Comment:
Just a comparison with what we do in the call tree.
In the call tree we also have such a map, but we populate it only incrementally, when the children for a specific node are asked for.
Here you build the whole cache at the first call. I'm concerned about the performance because in this algorithm you call `computeChildrenArray` a lot, each time this method is called the full zipFileTable is iterated. I believe we can get a freeze of the page if the Zip file is big enough.
I think it would be OK if you can iterate the zipFileTable once only.

Reply:
I made it per-call, and cached, rather than a complete upfront cost. Thanks for catching this.